### PR TITLE
[mkldnn-v1.0] Add MKL-DNN reshape&flatten&expand_dims

### PR DESCRIPTION
**Pull request by wuxun-zhang**
## Description ##
Add mkl-dnn 1.0 support for reshape/flatten/expand_dims ops.

@pengzhao-intel @ciyongch @TaoLv @ZhennanQin @rongzha1 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_reshape.cc [hunk 1]
```diff
@@ -89,42 +88,31 @@ MKLDNNReshapeFwd::MKLDNNReshapeFwd(const OpReqType &req,
...
+                         {MKLDNN_ARG_TO, *output.GetMKLDNNData()}});
+  } else if (prims_size == 2) {
+    auto temp_md = GetDesc(in_md, GetDefaultFormat(in_md));
+    temp_ = std::make_shared<mkldnn::memory>(temp_md, CpuEngine::Get()->get_engine(),
+              workspace);
+    args_map_.push_back({{MKLDNN_ARG_FROM, *in_mem},
+                        {MKLDNN_ARG_TO, *temp_}});
+    args_map_.push_back({{MKLDNN_ARG_FROM, *temp_},
+                        {MKLDNN_ARG_TO, *output.GetMKLDNNData()}});
+  }
```
Comment (rongzha1) [inline src/operator/nn/mkldnn/mkldnn_reshape.cc]: maybe add else { LOG(FATAL)<< "   "}  ?

Comment (wuxun-zhang) [reply src/operator/nn/mkldnn/mkldnn_reshape.cc]: Yes. Will try to add.

File: src/operator/nn/mkldnn/mkldnn_reshape.cc [hunk 2]
```diff
@@ -89,42 +88,31 @@ MKLDNNReshapeFwd::MKLDNNReshapeFwd(const OpReqType &req,
...
-  // register primitives
   auto stream = MKLDNNStream::Get();
-  for (auto &v : this->prims_) {
-    stream->RegisterPrim(v);
+  auto in_mem = input.GetMKLDNNData();
+  auto in_md = in_mem->get_desc();
+  // register primitives and arguments
+  size_t prims_size = prims_.size();
+  if (prims_size == 1) {
+    args_map_.push_back({{MKLDNN_ARG_FROM, *in_mem},
```
Comment (ZhennanQin) [inline src/operator/nn/mkldnn/mkldnn_reshape.cc]: Strange code. always push_back, but not clear? What will happen in 2nd execution?

Comment (wuxun-zhang) [reply src/operator/nn/mkldnn/mkldnn_reshape.cc]: Thanks for pointing this. I think I should put the definition of `args_map_` inside this function.

Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_reshape.cc]: Rename `args_map_` to `args_map`, it's not a class member.

File: src/operator/nn/mkldnn/mkldnn_reshape.cc [hunk 3]
```diff
@@ -48,39 +49,37 @@ MKLDNNReshapeFwd::MKLDNNReshapeFwd(const OpReqType &req,
...
-  auto temp_fmt = static_cast<mkldnn::memory::format>(GetDefaultFormat(in_pd.desc()));
-  auto temp_desc = mkldnn::memory::desc(temp_dims, temp_type, temp_fmt);
-  auto temp_pd = mkldnn::memory::primitive_desc(temp_desc, engine);
-  temp_ = std::make_shared<mkldnn::memory>(temp_pd, nullptr);
+  auto temp_md = GetDesc(in_md, GetDefaultFormat(in_md));
+  temp_ = std::make_shared<mkldnn::memory>(temp_md, engine);
 
   // destination
-  out_ = std::make_shared<mkldnn::memory>(temp_pd, nullptr);
+  out_ = std::make_shared<mkldnn::memory>(temp_md, engine);
```
Comment (ZhennanQin) [inline src/operator/nn/mkldnn/mkldnn_reshape.cc]: Why not create `out_` with nullptr?

File: src/operator/nn/mkldnn/mkldnn_reshape.cc [hunk 4]
```diff
@@ -89,42 +88,37 @@ MKLDNNReshapeFwd::MKLDNNReshapeFwd(const OpReqType &req,
...
+  auto in_md = in_mem->get_desc();
+  // register primitives and arguments
+  std::vector<mkldnn_args_map_t> args_map_;
+  size_t prims_size = prims_.size();
+  if (prims_size == 1) {
+    args_map_.push_back({{MKLDNN_ARG_FROM, *in_mem},
+                         {MKLDNN_ARG_TO, *output.GetMKLDNNData()}});
+  } else if (prims_size == 2) {
+    auto temp_md = GetDesc(in_md, GetDefaultFormat(in_md));
+    temp_ = std::make_shared<mkldnn::memory>(temp_md, CpuEngine::Get()->get_engine(),
```
Comment (ZhennanQin) [inline src/operator/nn/mkldnn/mkldnn_reshape.cc]: temp_ is already created when creating prims_, why we recreate this again?

File: src/operator/nn/mkldnn/mkldnn_reshape.cc (hunk 3 above)
Comment (wuxun-zhang) [reply src/operator/nn/mkldnn/mkldnn_reshape.cc]: Fixed

File: src/operator/nn/mkldnn/mkldnn_reshape.cc (hunk 4 above)
Comment (wuxun-zhang) [reply src/operator/nn/mkldnn/mkldnn_reshape.cc]: Yes, no need to create again. Fixed.

**Comment by pengzhao-intel** [comment]
please rebase the code and I will merge soon.

**Comment by wuxun-zhang** [comment]
This PR suffers the same onnx test error with [PR#16272](https://github.com/apache/incubator-mxnet/pull/16272#issuecomment-540282920). @TaoLv

**Comment by TaoLv** [comment]
Merging now.

**Comment by wuxun-zhang** [comment]
@ZhennanQin just filed a [PR](https://github.com/apache/incubator-mxnet/pull/16455) to address reshape issue, which is merged into master. We need to submit another PR to include this fix in mkldnn-v1.0 branch. I can try to rebase mkldnn-v1.0 branch from master and fix conflicts. FYI @TaoLv

**Pull request merged**